rfcs/0056-FE-APIs.md

```diff
@@ -0,0 +1,234 @@
+  addDocumentAction(actionFunction: DocumentActionFunction): void;
+}
+
+type DocumentActionFunction = (
+  previousActions: DocumentAction[],
+  ctx: EditViewContext
+) => DocumentAction[];
+
+interface DocumentAction {
+  label: string;
+  onClick: (
+    event: React.MouseEvent<HTMLButtonElement>,
+    ctx: EditViewContext
+  ) => void;
+  icon?: React.ElementType;
+  disabled?: boolean;
+  dialog?: DialogOptions | NotifiationOptions | ModalOptions;
+}
+
+interface DialogOptions {
+  type: "dialog";
+  title: string;
+  content?: React.ReactNode;
+  onConfirm?: () => void;
+  onCancel?: () => void;
+}
+
+interface NotifiationOptions {
+  type: "notifcation";
+  title: string;
+  link?: {
+    label: string;
+    url: string;
+    target?: string;
+  };
+  content?: string;
+  onClose?: () => void;
+  status?: "info" | "warning" | "softWarning" | "success";
+  timeout?: number;
+}
+
+interface ModalOptions {
+  type: "modal";
+  title: string;
+  content: React.ReactNode;
+  footer: React.ReactNode;
+  onClose?: () => void;
+}
+```
+
+This gives users the ability to optionally render a `Notification`, a `ConfirmationDialog` or a `Modal` depending on the complexity of the action they’re performing. Consider the “Publish Multiple Locales” action, this would be very complex and most likely use the `Modal` option.
+
+## Current InjectionZones
+
+The current injection zones affected by this RFC are:
+
+- `injectContentManagerComponent('editView', 'informations', conf)`
+
+We would deprecate this one with a clear guide on moving towards adding your own panel and/or adding a document action instead. As we progress we could deprecate other injection zone APIs with much clear configuration objects.
+
+# Tradeoffs
+
+It also stands to reason that perhaps that something like the `SideBar` would be better off not having things injected, but rather you simply overwrite the entire component and `renderDefaults` around it, the issue with this could be that plugins would overwrite the component at X opportunities so perhaps this is better left for a user’s Strapi Application configuration file instead.
+
+# Alternatives
+
+The alternative would be to use the current InjectionZones pattern more freely, but that has very little type safety and the pattern is essentially an uncontrolled closed-box where you “just pass props” and assume it goes to the right component leading to hard to test circumstances and unpredictability which imo makes for bad software. Product has also had concerns over the viability of the current InjectionZones pattern...
+
+# Unresolved questions
+
+- Do we want to allow DocumentActions to be written as if they’re react components but return a data structure? This allows actions to be even more powerful, but their implementation in Strapi becomes potentially more complex.
+- Should we live with injection zones in V4 and only deprecate them or completely remove them and do the additional work to define new APIs for the affected areas?
```

Comment:
Please do not remove injection zones, but allow a graceful upgrade to the new api.